feat: new import commands for dataset and databases

betodealmeida opened the pull request:
### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

This PR introduced 3 new commands:

- `superset.commands.importers.v1.ImportModelsCommand`: a base class used for the v1 importers for databases, datasets (in this PR), dashboard, charts and saved queries (in future PRs). The base class validates the import bundle, checking that the `metadata.yaml` file that looks like this:

```yaml
version: 1.0.0
type: Database
timestamp: '2020-11-04T21:27:44.423819+00:00'
```

Calling `run` on the class will start a database transaction block, calling the `import_bundle` method of each subclass. If there are any errors all imports are rolled back, preventing partial imports.

- `superset.databases.commands.importers.v1.ImportDatabasesCommand`: a command that imports one or more databases present in the import bundle, as well as associated datasets. Because of this, in addition to validating databases present in the bundle, the command will also validate the datasets.

**When importing a database, associated datasets are imported non-destructively**: if a table `foo` currently has a column `col`, but the imported dataset doesn't have it the column **will not be deleted**.

- `superset.datasets.commands.importers.v1.ImportDatasetsCommand`: a command that imports one or more datasets present in the import bundle, as well as associated databases. Similar to the previous command, it will validate the database configuration as well.

When importing a dataset we **sync** columns and metrics one way (from bundle to DB). This means that when importing an existing dataset all missing columns and metrics in the import bundle are removed. This behavior is similar to the current datasource CLI import in Superset when passing the `--sync` flag.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

N/A

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

Added unit tests covering import, multiple imports, validation, and rollback in case of error, ensuring that the imports are atomic and idempotent.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [X] Has associated issue: https://github.com/apache/incubator-superset/issues/11167
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

betodealmeida committed:
feat: commands for importing databases and datasets

hughhhh commented on superset/commands/importers/v1.py:
```diff
@@ -0,0 +1,113 @@
+from superset.models.helpers import ImportExportMixin
+
+METADATA_FILE_NAME = "metadata.yaml"
+IMPORT_VERSION = "1.0.0"
```
do you have a process on when this would be upgraded to 1.0.1 vs. 2.0.0

betodealmeida replied:
The idea is that if we bump the format to, say, `1.0.1` we would change the validation to accept anything between `1.0.0` and `1.0.1`, and adjust the class to handle the differences. If we bump it to `2.0.0` we should right a new command under `v2`, so that we can still be able to import `1.0.0` with the `v1` command.

hughhhh commented on superset/commands/importers/v1.py:
```diff
@@ -0,0 +1,113 @@
+    def import_(
+        session: Session, config: Dict[str, Any], overwrite: bool = False,
+    ) -> ImportExportMixin:
+        raise NotImplementedError("Subclasss MUST implement import_")
```
```suggestion
	raise NotImplementedError("Subclass MUST implement import_")
```

betodealmeida replied:
Argh, good catch! We need a spellchecker pre-commit hook.

hughhhh commented on superset/datasets/commands/importers/v1.py:
```diff
@@ -0,0 +1,136 @@
+        session: Session, config: Dict[str, Any], overwrite: bool = False,
+    ) -> SqlaTable:
+        existing = session.query(SqlaTable).filter_by(uuid=config["uuid"]).first()
+        if existing:
```
couldn't you go `if existing and not overwrite`

betodealmeida replied:
It needs to be treated separately. If the imported table already exists there are 2 scenarios:

- If `overwrite` is false them we shouldn't do anything, so `return`
- if `overwrite` is true then we can continue, but first we need to set the current ID to replace insert of insert

So if we could write it:

```python
if existing and not overwrite:
    return

if existing:
    config["id"] = existing.id
```

Which is a bit more verbose.
```

hughhhh commented on superset/commands/importers/v1.py:
```diff
@@ -0,0 +1,113 @@
+                f'Missing file "{METADATA_FILE_NAME}" in contents'
+            )
+        content = self.contents[METADATA_FILE_NAME]
+        metadata = load_yaml(METADATA_FILE_NAME, content)
```
nit:
```suggestion
metadata = load_yaml(METADATA_FILE_NAME, self.contents[METADATA_FILE_NAME])
```
and remove the top line

betodealmeida committed:
Refactor code

betodealmeida commented:
Based on personal code review with @willbarrett:

- got rid of inheritance
- simplified the logic of the commands leaving only 1 private method
- removed circular imports
